### PR TITLE
Rescan auction for existing bids

### LINKED ISSUE #108: Auctions should display all bids, not just the current user's
Currently, the auction screen displays only bids placed by the wallet user. We should change this so that all bids are displayed in a similar manner to Namebase.

### LINKED ISSUE #113: Update HSD
Use the version that includes the patch for invalid post-import balances.

### PR DESCRIPTION
**Pull request by pinheadmz**
Closes #108 
Closes #113 

### Changes Summary

- Import branch of hsd with `rpc importname` (https://github.com/handshake-org/hsd/pull/408). We can wait until this is merged into master and rebase, but this branch is already based on latest hsd master.
- Expose wallet `rpc getwalletinfo` (as opposed to wallet HTTP `getinfo`), and `rpc importname` in background services.
- Add a few props and actions to redux store to start, stop, and monitor rescan progress.
- Introduce new modal that blocks the UI while the wallet is syncing. Progress is determined by comparing walletDB height with blockchain height. Modal applied to:
  - Importing seed and rescanning entire chain
  - Rescanning an in-progress auction to discover existing bids
- Auction rescanning is triggered by one of two actions:
  - Manually clicking "Rescan Auction" button on the auction page, _before_ a user has entered the auction (discover all existing bids _before_ a user places their first bid)
  - Automatically rescan auction when a user places a bid, if they have not already imported the name. 
- Minor:
  - Pass `left` prop to tooltip to keep hover-popups on the screen if they are too far to the right.
  - Add `walletHasName` to `domain` object, indicating the name is stored in the hsd wallet (which is separate from the Bob "watchlist" -- although they could be merged in the future)

### Potential issues

- When the user places a bid on a new auction, Bob will import the name and rescan the auction _before_ placing the bid. This is necessary because `importname` will throw if the name is already stored in the wallet, which it would be if the bid was sent first. This could be modified but I had issues finding existing bids after sending a bid, will investigate further.
- Rescans aren't dangerous (compared to a seed-importing-restore) but they do take a few seconds and might confuse the user.
- The WalletSync modal stays up until wallet sync is 100%. This should always happen eventually but there is no way to close the modal if something goes wrong. However, this would indicate a much bigger issue so maybe it's best to be "frozen" instead of the user thinking everything is OK.

### Rescan Auction button with hovering tooltip

rescan-auciton-button

### Manually triggering auction rescan to load existing bids

rescan-auction

### Automatic auction rescan when placing first bid

This example bidding on a mainnet name auction just before close to give an upper-bound example on the amount of blocks to rescan.

bid-auction-rescan

### CODE REVIEW

**Comment by pinheadmz** [comment]
rebase to af995a7b1a0a2fd9b00cb791606647f2aa3f5a0f:

Rebase on master (#116  just merged)
Update changelog

File: app/pages/Auction/BidActionPanel/BidNow.js [hunk 1]
```diff
@@ -28,6 +30,8 @@ class BidNow extends Component {
     ownHighestBid: PropTypes.object.isRequired,
     isPending: PropTypes.bool.isRequired,
     confirmedBalance: PropTypes.number.isRequired,
+    getNameInfo: PropTypes.func.isRequired,
```
Comment (mslipper) [inline app/pages/Auction/BidActionPanel/BidNow.js:33]: Assuming we don't refactor the logic in `sendBid` into the wallet service, it'd be cleaner if `getNameInfo` et. al. were all passed in via `connect`. I see that `BidActionPanel` is receiving `getNameInfo` via view props, which seems tough from a maintainability standpoint given that every instance of `BidActionPanel` will need `getNameInfo`.

File: app/pages/Auction/BidActionPanel/BidNow.js [hunk 2]
```diff
@@ -59,11 +63,19 @@ class BidNow extends Component {
...
+    const {sendBid, domain} = this.props;
     const {bidAmount, disguiseAmount} = this.state;
     const lockup = Number(disguiseAmount) + Number(bidAmount);
 
     try {
+      // If the name is not already stored in the wallet, rescan the auction
+      // for past bids before proceeding. The wallet will continue to track
+      // newer bids automatically.
+      if (!domain.walletHasName) {
+        this.props.waitForWalletSync();
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply app/pages/Auction/BidActionPanel/BidNow.js:75]: Yeah its funky because we want the wallet progress modal to run asynchronously in "parallel" with the actual wallet import-and-rescan process. I played with a few combinations of call sequences and awaits and this was the best UX (modal appears and stays up until rescan is done)

Although looking at it again it looks like an edge case could exist where the progress modal will load, quickly hit 100% and exit before the wallet import-and-rescan has even started, but I haven't seen that in any of my tests.

The sequence is a little different when restoring from seed. This was required because the import seed function actually restarts the wallet service completely, so the progress monitor must be called last:

https://github.com/pinheadmz/bob-wallet/blob/af995a7b1a0a2fd9b00cb791606647f2aa3f5a0f/app/pages/Onboarding/ImportSeedFlow/index.js#L118-L120

File: app/pages/Auction/BidActionPanel/BidNow.js [hunk 3]
```diff
@@ -76,9 +88,18 @@ class BidNow extends Component {
       console.error(e);
       logger.error(`Error received from BidNow - sendBid]\n\n${e.message}\n${e.stack}\n`);
       this.props.showError('Failed to place bid. Please try again.');
+    } finally {
+      await this.props.getNameInfo(domain.name);
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply app/pages/Auction/BidActionPanel/BidNow.js:92]: I'm using `getNameInfo()` here to force a refresh of the UI. I can wrap it in an error-catcher as well... but this is making me wonder if maybe we should add the error notifications in the `walletClient` itself? Then perhaps we wouldn't have to wrap each individual call throughout the app.

File: app/pages/Auction/BidActionPanel/BidNow.js [hunk 4]
```diff
@@ -59,11 +63,19 @@ class BidNow extends Component {
...
-    const {sendBid} = this.props;
+    const {sendBid, domain} = this.props;
     const {bidAmount, disguiseAmount} = this.state;
     const lockup = Number(disguiseAmount) + Number(bidAmount);
 
     try {
+      // If the name is not already stored in the wallet, rescan the auction
+      // for past bids before proceeding. The wallet will continue to track
+      // newer bids automatically.
+      if (!domain.walletHasName) {
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply app/pages/Auction/BidActionPanel/BidNow.js:74]: Ah, good point. Well, it will only `importName` if this is the first time you bid on an auction. And if you opened the auction it won't happen. I'm imagining that users take advantage of the "rescan auction" button before placing their first bid. This just shims it in anyway if they don't.

I'll look into this approach though, makes sense.

File: app/pages/Auction/BidActionPanel/BidNow.js (hunk 1 above)
Comment (pinheadmz) [reply app/pages/Auction/BidActionPanel/BidNow.js:33]: Ah good point here too, will look into it.

File: app/ducks/walletActions.js [hunk 5]
```diff
@@ -106,6 +109,25 @@ export const send = (to, amount, fee) => async (dispatch) => {
   return res;
 };
 
+export const waitForWalletSync = () => async (dispatch, getState) => {
+  await dispatch({type: START_SYNC_WALLET});
+
+  for (;;) {
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply app/ducks/walletActions.js:117]: Fixed in https://github.com/kyokan/bob-wallet/pull/117/commits/c26e9aeb419d323d6f8b3a8625f49ba79caa81c9

I set the timeout for 5 seconds (5 loops with no change in progress)

File: app/pages/Auction/BidActionPanel/BidNow.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (pinheadmz) [reply app/pages/Auction/BidActionPanel/BidNow.js]: see comment below

File: app/pages/Auction/BidActionPanel/BidNow.js [hunk 6]
```diff
@@ -76,9 +88,18 @@ class BidNow extends Component {
...
       logger.error(`Error received from BidNow - sendBid]\n\n${e.message}\n${e.stack}\n`);
       this.props.showError('Failed to place bid. Please try again.');
+    } finally {
+      await this.props.getNameInfo(domain.name);  
     }
   };
 
+  rescanAuction = async () => {
+    const {domain} = this.props;
+    this.props.waitForWalletSync();
```
[the remark this answers is not in the document]
Comment (pinheadmz) [reply app/pages/Auction/BidActionPanel/BidNow.js]: see comment below

**Comment by pinheadmz** [comment]
The async, sync progress timeout, and error handling stuff is addressed in https://github.com/kyokan/bob-wallet/pull/117/commits/1dcdd8ced222ae98830ad32cefe2020ce720c559 and is a bit difficult to test but here's how I ran through it:

There's three functions affected by this PR:
- Rescan Auction (which calls `importname` then rescans)
- Bid on auction (auto-triggering a rescan auction before placing the bid)
- Import seed into new wallet (auto-triggering a wallet rescan and displaying sync progress modal)

There's two errors hsd/rpc may throw that we need to catch and show the user:
- hsd walletDB `rescanBlock`
- hsd wallet rpc `importname`
- (third case: no errors are thrown)

Using Bob as a full node (in regtest), I connected a local SPV node to it to send external bids and auction opens. I added bugs/errors systematically in the two hsd functions above and ran through all three Bob functions to test the UX:

### rpc `importname` throws
- rescan auction
  - sync progress modal flashes for a frame
  - rpc error displayed in notification
  - after reload, wallet state is unchanged
  - summary: if `importname` throws, error is displayed and nothing happens

- place bid
  - sync progress modal flashes for a frame
  - rpc error is displayed
  - bid SUCCESS modal is displayed _on top of error notification_
  - mine 1 block, confirming bid
  - summary: if `importname` throws, the `sendbid` will still succeed but the wallet will not "know" about any previous bids in the auction. It will display its own bid and will continue displaying bids placed in the future.

### `rescanBlock` throws
- rescan auction
  - sync progress modal displays for 5 seconds then displays "sync stalled" error notification. The actual error is printed to hsd log.
  - after reload, wallet state is unchanged
  - summary: if `rescanBlock` throws, error is displayed and nothing happens
- place bid
  - sync modal displays for 5 seconds then displays concatenated error "failed to place bid: sync progress stalled"
  - `sendbid` FAILS to send bid.
  - after reload, wallet state is unchanged
  - summary: if the wallet can not rescan the auction, it will not send the bid
- import seed
  - sync modal displays for 5 seconds then displays "sync stalled" error.
  - wallet state is fucked, but there is no indication of this to the user. (assuming the rescan error was not a permanent bug like I introduced for this test, the next new block added to chain will trigger a walletDB "catch up")

### No hsd errors
- startup
  - if wallet state was previously out of sync, wallet will automatically rescan and catch up to chain (no indication to user that wallet is catching up, this is a background hsd process)
- rescan auction
  - sync modal runs to 100%, then displays all previous bids. All future bids are displayed from inside or outside the wallet
- place bid
  - sync modal runs to 100%, followed by BID SUCCESS modal. All previous bids are displayed and all future bids are displayed (after confirmation)
- import seed and create new wallet
  - after confirming "opt in" option, sync progress modal appears and runs to 100%, then returns to wallet home screen with wallet state completely intact.


### Additional notes (TODO in future work):

- The UI does not update automatically when a bid is confirmed. A manual reload (cntrl+R) is required to see the wallet's own bid displayed on the auction page.

- Add a wallet status indicator in the bottom left corner next to chain height to indicate that walletDB is in sync with chain, maybe disable wallet actions in this state.

- Maybe add pending bids to the "my bids" table on the auction screen.

File: app/pages/Auction/BidActionPanel/BidNow.js (hunk 1 above)
Comment (pinheadmz) [reply app/pages/Auction/BidActionPanel/BidNow.js:33]: Wow ok sorry about this mess. Got confused by the redux structure. Fixed in e84f7de0731f4cf3b171ab9032cee23092f5081b and also fixed the same thing I did in https://github.com/kyokan/bob-wallet/pull/116

File: app/pages/Auction/BidActionPanel/BidNow.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (pinheadmz) [reply app/pages/Auction/BidActionPanel/BidNow.js]: Fixed in d2eff90ced3d8e66eba44d0e9f08f7e18112945c

**Comment by pinheadmz** [comment]
rebase to  2d2bb2416c820adde46af1456b3d3af2c190dd98 :
rebase on master, fix conflicts.

File: app/ducks/walletActions.js [hunk 7]
```diff
@@ -106,6 +109,54 @@ export const send = (to, amount, fee) => async (dispatch) => {
...
+      stall++;
+    } else {
+      lastProgress = progress;
+      stall = 0;
+    }
+
+    if (stall >= 5) {
+      dispatch({type: STOP_SYNC_WALLET});
+      throw new Error('Wallet sync progress has stalled.');
+      break;
```
Comment (mslipper) [inline app/ducks/walletActions.js:146]: No need to break if there's an error thrown :).
```suggestion
```
